src/compiler/utilities.ts

```diff
@@ -6967,6 +6967,49 @@ namespace ts {
         return compilerOptions[flag] === undefined ? !!compilerOptions.strict : !!compilerOptions[flag];
     }
 
+    export function compilerOptionsAffectSemanticDiagnostics(newOptions: CompilerOptions, oldOptions: CompilerOptions) {
+        if (oldOptions === newOptions) {
+            return false;
+        }
+
+        return changedCompileOptionValueOf(newOptions, oldOptions, [
+            "noImplicitReturns",

```

Comment:
I wonder if, instead of placing the list here, we use our central list of compiler options and mark them all with an internal field called "semantic" or something

Reply:
That sounds good, but we still have these strict flags which should check their total value (from individual flag and strict flag) to see if it has really changed so we would need that part either ways.